Possible bug in /pkg/engine/common/base.go%Do()

### Discussed in https://github.com/projectdiscovery/katana/discussions/1394

<div type='discussions-op-text'>

<sup>Originally posted by **mjkim610** September  5, 2025</sup>
I think I have come across a bug where SIGSEGV occurs because of an attempt to access `resp.Resp` when both `err` and `resp` are nil at `base.go%Do()`.

If this issue is confirmed as a bug, I will create an issue and create a fix PR for it.

## How to reproduce:

cmd:
```
❯ go build ; ./katana -u https://brokencrystals.com/ -headless -depth 2 -v
```

log trace:
```
   __        __                
  / /_____ _/ /____ ____  ___ _
 /  '_/ _  / __/ _  / _ \/ _  /
/_/\_\\_,_/\__/\_,_/_//_/\_,_/                                                   

                projectdiscovery.io

[INF] Current katana version v1.2.2 (latest)
[INF] Started headless crawling for => https://brokencrystals.com/
[...]
[GET] https://brokencrystals.com/graphiql/main.js
[WRN] mjkim - resp itself is null! // log added to verify source of crash
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1050f19b0]

goroutine 5601 [running]:
github.com/projectdiscovery/katana/pkg/engine/common.(*Shared).Do.func1()
        [...]/katana/pkg/engine/common/base.go:305 +0x120
created by github.com/projectdiscovery/katana/pkg/engine/common.(*Shared).Do in goroutine 40
        [...]/katana/pkg/engine/common/base.go:274 +0x234
```</div>

--- a/pkg/engine/common/base.go
+++ b/pkg/engine/common/base.go
@@ -298,7 +298,7 @@ func (s *Shared) Do(crawlSession *CrawlSession, doRequest DoRequestFunc) error {
 				_ = s.Options.OutputWriter.WriteErr(outputError)
 				return
 			}
-			if resp.Resp == nil || resp.Reader == nil {
+			if resp == nil || resp.Resp == nil || resp.Reader == nil {
 				return
 			}
 			if s.Options.Options.DisableRedirects && resp.IsRedirect() {

--- a/pkg/engine/hybrid/crawl.go
+++ b/pkg/engine/hybrid/crawl.go
@@ -229,7 +229,7 @@ func (c *Crawler) navigateRequest(s *common.CrawlSession, request *navigation.Re
 		return nil, errkit.Wrap(err, "hybrid: url could not be parsed")
 	}
 
-	if response.Resp == nil {
+	if response == nil || response.Resp == nil {
 		return nil, errkit.Wrap(err, "hybrid: response is nil")
 	}
 	response.Resp.Request.URL = parsed.URL
@@ -245,9 +245,11 @@ func (c *Crawler) navigateRequest(s *common.CrawlSession, request *navigation.Re
 	}
 
 	response.Body = body
-	response.Reader.Url, _ = url.Parse(request.URL)
-	if c.Options.Options.FormExtraction {
-		response.Forms = append(response.Forms, utils.ParseFormFields(response.Reader)...)
+	if response.Reader != nil {
+		response.Reader.Url, _ = url.Parse(request.URL)
+		if c.Options.Options.FormExtraction {
+			response.Forms = append(response.Forms, utils.ParseFormFields(response.Reader)...)
+		}
 	}
 
 	response.Reader, err = goquery.NewDocumentFromReader(strings.NewReader(response.Body))

